bbl_screen-patch/patches/printerui/qml/x1plus/GpioActions.js

```diff
@@ -0,0 +1,192 @@
+    console.log("[x1p] gpio button settings restored to defaults.");
+    return defaultsApplied; 
+}
+
+/**
+ * Updates the action and parameters for a given button's press type.
+ * @param {string} buttonName - The name of the button ("cfw_power" or "cfw_estop").
+ * @param {string} pressType - The type of press ("shortPress" or "longPress").
+ * @param {number} actionVal - The value of the action to set.
+ * @param {Object} _parameters - Parameters for the action.
+ */
+function updateButtonAction(buttonName, pressType, actionVal, _parameters = {}) {
+    var _gpio = gpioBinding();
+    if (!_gpio[buttonName] || !_gpio[buttonName][pressType]) {
+        console.log(`[x1p] Invalid or missing configuration for button: ${buttonName}, pressType: ${pressType}`);
+        return; 
+    }
+    let actionObject = buttonActions.find(a => a.val === actionVal);
+    if (!actionObject) {
+        console.log(`[x1p] Action value ${actionVal} not found in buttonActions.`);
+        return;
+    }
+    _gpio[buttonName][pressType].action = actionVal; 
+    _gpio[buttonName][pressType].parameters = _parameters;
+    _gpio[buttonName][pressType].default = getDefault(buttonName,pressType);
+   
+    _setGpio(_gpio);
+    X1Plus.atomicWriteJson(settings_file, _gpio);
+}
+
+
+
+/**
+ * Retrieves the action text for a given button type and press type.
+ * @param {string} buttonName - The name of the button ("cfw_power" or "cfw_estop").
+ * @param {string} pressType - The type of press ("shortPress" or "longPress").
+ * @return {string} - The action text.
+ */
+function getActionText(buttonName, pressType) {
+    if (!buttonName || !pressType) {
+        return "";
+    }
+    var _gpio = gpioBinding();
+    if (_gpio[buttonName] && _gpio[buttonName][pressType]) {
+        let actionVal = _gpio[buttonName][pressType].action;
+        let actionObject = buttonActions.find(a => a.val === actionVal);
+        if (actionObject) {
+            console.log(`Action text for ${buttonName} ${pressType}: ${actionObject.name}`);
+            return actionObject.name;
+        } else {
+            console.log(`Action value ${actionVal} not found in buttonActions for ${buttonName} ${pressType}.`);
+        }
+    } else {
+        console.log(`[x1p] Action not mapped for ${buttonName}, pressType: ${pressType}.`);
+    }
+    return "";
+}
+
+
+function _loadSettings(noload) {
+    if (noload) return;
+    let loadedSettings = X1Plus.loadJson(settings_file);
+    if (loadedSettings && Object.keys(loadedSettings).length > 0) {
+        console.log("[x1p] Loaded settings:", JSON.stringify(loadedSettings));
+        _setGpio(loadedSettings);
+    } else {
+        console.log("[x1p] no settings file found. loading default actions.");
+        resetToDefaultActions(); 
+    }
+}
+
+/**
+ * handle button action given dds message
+ * @param {string} action_code
+ * @param {string} datum
+ */
+function handleAction(action_code, datum) {
```

Comment:
also I wonder if this should just live in dds, rather than here.  GpioActions is for configuring how the python script will run, and in theory, many user macros might want to trigger QML events, not just key events?

Reply:
I like that idea, and I'm wondering should we just do all event handling in QML? Unless there's a benefit to running Python macros from gpiokeys.py, that can just be done from QML and a chunk of gpiokeys.py can be simplified?